fem/tbilininteg.hpp

```diff
@@ -554,6 +554,1126 @@ struct TDiffusionKernel<3,3,complex_t>
+            R.val_qpts(i,1,j,k) = A21*x1 + A22*x2 + A32*x3;
+            R.val_qpts(i,2,j,k) = A31*x1 + A32*x2 + A33*x3;
+         }
+      }
+   }
+};
+
+// mass kernel for RT Finite Element
+
+// complex_t - type for the assembled data
+template <int SDim, int Dim, typename complex_t>
+struct THDivDivKernel;
+
+// mass kernel for RT Finite Element kernel in 2D
+template <typename complex_t>
+struct THDivDivKernel<2,2,complex_t>
+{
+   typedef complex_t complex_type;
+
+   // needed for the TElementTransformation::Result class
+   static const bool uses_Jacobians = true;
+
+   // needed for the FieldEvaluator::Data class
+   static const bool in_values  = false;
+   static const bool in_divs    = true;
+   static const bool out_values = false;
+   static const bool out_divs   = true;
+
+   // Partially assembled data type for one element with the given number of
+   // quadrature points. This type is used in partial assembly, and partially
+   // assembled action. Stores one symmetric 1 x 1 matrix per point.
+   template <int qpts>
+   struct p_asm_data { typedef TVector<qpts,complex_t> type; };
+
+   // Partially assembled data type for one element with the given number of
+   // quadrature points. This type is used in full element matrix assembly.
+   // Stores one general (non-symmetric) 1 x 1 matrix per point.
+   template <int qpts>
+   struct f_asm_data { typedef TVector<qpts,complex_t> type; };
+
+   template <typename IR, typename coeff_t, int NE>
+   struct CoefficientEval
+   {
+      typedef typename IntRuleCoefficient<IR,coeff_t,NE>::Type Type;
+   };
+
+   // Method used for un-assembled (matrix free) action.
+   // Jt        [M x Dim x SDim x NE] - Jacobian transposed, data member in F
+   // Q                               - CoefficientEval<>::Type
+   // q                               - CoefficientEval<>::Type::result_t
+   // val_qpts [M x SDim x NC x NE]  - in/out data member in R
+   //
+   // curl_qpts = (w/det(J))J val_qpts
+   template <typename T_result_t, typename Q_t, typename q_t,
+             typename S_data_t>
+   static inline MFEM_ALWAYS_INLINE
+   void Action(const int k, const T_result_t &F,
+               const Q_t &Q, const q_t &q, S_data_t &R)
+   {
+      const int M = S_data_t::eval_type::qpts;
+      const int NC = S_data_t::eval_type::vdim;
+      MFEM_STATIC_ASSERT(T_result_t::Jt_type::layout_type::dim_1 == M,
+                         "incompatible dimensions");
+      MFEM_FLOPS_ADD(M*(4+NC));
+      for (int i = 0; i < M; i++)
+      {
+         typedef typename T_result_t::Jt_type::data_type real_t;
+         const real_t J11 = F.Jt(i,0,0,k);
+         const real_t J12 = F.Jt(i,1,0,k);
+         const real_t J21 = F.Jt(i,0,1,k);
+         const real_t J22 = F.Jt(i,1,1,k);
+         const complex_t w_det_J = Q.get(q,i,k) / (J11 * J22 - J21 * J12);
+         for (int j = 0; j < NC; j++)
+         {
+            const complex_t x1 = R.div_qpts(i,0,j,k);
```

Comment:
Does the divergence needs to be stored? Why is it not computed on the fly with a tensor?

Reply:
Does this operator works for non-H(div) basis? The operator should be able to compute the divergence for a H1 basis too, no?

Reply:
No, it is design for H(div) space. The main reason is that the mapping for different spaces are not the same.